src/core/cgroup.c

```diff
@@ -774,13 +774,58 @@ void cgroup_oomd_xattr_apply(Unit *u, const char *cgroup_path) {
                 unit_remove_xattr_graceful(u, cgroup_path, "user.oomd_omit");
 }
 
+int cgroup_log_xattr_apply(Unit *u, const char *cgroup_path) {
+        ExecContext *c;
+        char *pattern;
+        /* Use _cleanup_free_ instead of _cleanup_strv_free_ because we need to clean the strv only, not
+         * the strings owned by ExecContext. */
+        _cleanup_free_ char **patterns_strv = NULL;
+        _cleanup_free_ char *patterns_joined = NULL;
+        size_t len;
+        int r;
+
+        assert(u);
+
+        c = unit_get_exec_context(u);
+        if (!c)
+                return log_debug_errno(-EINVAL, "Failed to get ExecContext from Unit: %m");
+
+        if (set_isempty(c->log_filter_allowed_patterns) && set_isempty(c->log_filter_denied_patterns)) {
+                unit_remove_xattr_graceful(u, cgroup_path, "user.journald_log_filter_patterns");
+                return 0;
+        }
+
+        patterns_strv = set_get_strv(c->log_filter_allowed_patterns);
+        if (!patterns_strv)
+                return log_oom();
+
+        r = strv_push(&patterns_strv, (char *)"\xff");
+        if (r < 0)
+                return log_debug_errno(r, "Failed to push to strv: %m");
+
+        SET_FOREACH(pattern, c->log_filter_denied_patterns) {
+                r = strv_push(&patterns_strv, pattern);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to push to strv: %m");
+        }
+
+        r = strv_make_nulstr(patterns_strv, &patterns_joined, &len);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get nulstr from strv: %m");
```

Comment:
hmm, if you serialize things like this, there will be `0x00 0xFF 0x00` in the middle. I think it should be `0x00 0xFF` only. i.e. please convert both parts to nulstr, and then copy them together, separating them by 0xFF.